File: Week_02/cpp/src/_530_MinimumAbsoluteDifferenceInBst.cpp
#include <iostream>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution
{
public:
    int getMinimumDifference(TreeNode *root)
    {
        int ret = INT_MAX;
        TreeNode *pre = NULL;
        while (root)
        {
            if (!root->left)
            {
                calculate(&pre, &ret, root);
                root = root->right;
            }
            else
            {
                auto p = root->left;
                while (p->right && p->right != root)
                {
                    p = p->right;
                }
                if (!p->right)
                {
                    p->right = root;
                    root = root->left;
                }
                else
                {
                    p->right = NULL;
                    calculate(&pre, &ret, root);
                    root = root->right;
                }
            }
        }
        return ret;
    }
    void calculate(TreeNode **pre, int *ret, TreeNode *node)
    {
        if ((*pre))
        {
            *ret = min(*ret, abs(node->val - (*pre)->val));
        }
        (*pre) = node;
    }
};

File: Week_02/kotlin/src/_530_MinimumAbsoluteDifferenceInBst.kt
import java.util.*
import kotlin.math.abs

class _530_Dfs_Iterative {
    fun getMinimumDifference(root: TreeNode?): Int {
        val stack = LinkedList<TreeNode>()
        var node = root
        var ret = Integer.MAX_VALUE
        var pre: TreeNode? = null
        while (node != null || stack.isNotEmpty()) {
            while (node != null) {
                stack.push(node)
                node = node.left
            }
            node = stack.pop()
            if (pre != null) {
                ret = minOf(ret, abs(pre.`val` - node!!.`val`))
            }
            pre = node
            node = node.right
        }
        return ret
    }
}